Finished Exercise 4

yargav opened the pull request.

yargav committed:
Fetch Data From Network

yargav committed:
Move Data Fetching to FeedRepository

Still bug in where to change load at with this change..

yargav committed:
Add Initial Database Code

Only creation of classes, no actual database used yet

yargav committed:
Add Working Repository (From Network and Database)

yargav committed:
Inject Dependencies in Repository

Dependencies injected manually: Feed Data Api and Feed Database

yargav committed:
Add Tests For Feed Repository

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,24 @@
+
+@Dao
+interface FeedItemDao {
+    @Query("SELECT * FROM feedItems")
```
The table name in the Entity class is `FeedItems` (capital "F"). 
And here it's `feedItems` (lowercase "f").
This is OK since SQLite is case insensitive. 
But it can be confusing to somebody else reading the code.
The convention in SQL is is to use underscores to separate words (e.g. `feed_items`).

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/data/FeedRepository.kt:
```diff
@@ -1,9 +1,53 @@
+            .subscribe()
+    }
+
+    fun refresh(): Completable {
```
If somebody will call `refresh` from the main thread it will crash.
You should either add `.subscribeOn(Schedulers.io())` here or you should mention it in the documentation.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/database/DataAccessObjects.kt:
```diff
@@ -1,5 +1,24 @@
+    @Query("SELECT COUNT(*) FROM feedItems")
+    fun getCount(): Int
+
+    @Insert(onConflict = OnConflictStrategy.IGNORE)
```
I think it makes more sense to override the record. In the end we want to be able to refresh the feed.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:
```diff
@@ -1,5 +1,26 @@
+data class FeedData(@Json(name = "templatesMetadata") val metaData: List<Template> = listOf()){
+    @JsonClass(generateAdapter = true)
+    data class Template(
+        @Json(name = "configuration") val configuration: String,
```
The @Json() annotations are not necessary if the properties have the same name.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/network/DataTransferObjects.kt:
```diff
@@ -1,5 +1,26 @@
+@JsonClass(generateAdapter = true)
+data class FeedData(@Json(name = "templatesMetadata") val metaData: List<Template> = listOf()){
+    @JsonClass(generateAdapter = true)
+    data class Template(
```
I think that having a nested class will prevent us from reusing `Template` in other network requests in the future.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
     private val isEmpty = MutableLiveData<Boolean>()
     private val feedItems = MediatorLiveData<List<FeedItem>>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private val feedRepository = FeedRepository(
```
`FeedRepository` is supposed to have a different lifecycle than `FeedViewModel`. If we will add more `Fragments` to the app we don't want to recreate the repository for each screen. 

In other words we want to inject `FeedRepository` into  `FeedViewModel`. And the simplest way to do it is probably using `FeedViewModelFactory`.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
     private val feedItems = MediatorLiveData<List<FeedItem>>()
     private val networkErrorEvent = MutableLiveData<Event<String>>()
+    private val feedRepository = FeedRepository(
+        Room.databaseBuilder(
```
We also want to make the database a Singleton.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
+            application.applicationContext,
+            FeedDatabase::class.java, "FeedDatabase"
+        ).build(),
+        Retrofit.Builder().baseUrl(BASE_URL)
```
`Retrofit` should also be a Singleton.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
+        isEmpty.postValue(false)
+        //TODO change temporary load solution
+        isLoading.postValue(true)
+        Thread.sleep(10)
```
Please, delete `Thread.sleep(10)`.

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -1,15 +1,140 @@
+
+    @Test
+    fun feedApiService_InitializationTest() {
+        val feedData = mockFeedApiService.getFeedData().blockingGet()
```
Looks like this test doesn't test `FeedRepository` and doesn't belong here.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
+        //TODO change temporary load solution
+        isLoading.postValue(true)
+        Thread.sleep(10)
+        isLoading.postValue(false)
```
`refresh` should call `feedRepository.refresh()`.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -1,46 +1,87 @@
+        isLoading.postValue(true)
+        Thread.sleep(10)
+        isLoading.postValue(false)
+//        feedItems.postValue(feedItems.value?)
```
Please, handle `networkErrorEvent` that comes from `feedRepository`.

bmax-ltx reviewed (changes requested):
First of all, good job!
I see that you understand the concepts well.

yargav committed:
Add Database Changes

Name changed to go along with SQL naming convention.
Policy for on conflict (in the insert all function) changed from
ignore to override.

yargav committed:
Add Network Changes

Getting rid of unnecessary @json annotations.
Take the class "Template" out of FeedData (no longer a nested class).

yargav committed:
Change Refresh Methods

In Feed Repository, allow refresh to be called from main thread.
Call Feed Repository refresh from FeedViewModel's refresh.

yargav committed:
Add Some View Model Changes

Changes include using the feed view model factory to create an
instance of the view model that has a dependency injection of the
feed repository and making the network fetcher / database singletons.

yargav committed:
Remove Irrelevant Test

Still to do: deal with network error, deal with communication with
database and repository (why the feed items isn't updated after
templates removed from database).

yargav committed:
Handle Network Error

yargav committed:
Update Tests

bmax-ltx commented on app/src/androidTest/java/com/lightricks/feedexercise/data/FeedRepositoryTest.kt:
```diff
@@ -1,15 +1,116 @@
+    @Test
+    fun refresh_SavedItemsInDatabaseTest() {
+        feedRepository.refresh().blockingAwait()
+        assertThat(feedDatabase.feedItemDao().getAll().blockingObserve()).isEqualTo(
```
👍

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -20,27 +25,49 @@ open class FeedViewModel : ViewModel() {
+                isEmpty.postValue(false)
+                feedItems.postValue(result)
+            }
+            isLoading.postValue(false)
```
The actual loading is done in `refresh()`, so I think you should set `isLoading` over there.

bmax-ltx commented on app/src/main/java/com/lightricks/feedexercise/ui/feed/FeedViewModel.kt:
```diff
@@ -20,27 +25,49 @@ open class FeedViewModel : ViewModel() {
+    }
+
+    private fun handleError(message: String?){
+        networkErrorEvent.postValue(Event(message?:"Error without message"))
```
```suggestion
        networkErrorEvent.postValue(Event(message ?: "User friendly error message"))
```

bmax-ltx reviewed (approved):
Looks good, thanks!

yargav committed:
Change Error Message and Update Loading Location
